File: data_structures_and_algorithms/Code_401/class-06/solutions/java/LinkedList.java
import java.util.*;

class LinkedList <T> {
  public static void main(String[] args) throws Exception{
    LinkedList<Integer> linky = new LinkedList<>();
    linky.append(1);
    linky.append(3);
    linky.append(5);
    System.out.println(linky);
    linky.insertAfter(2, 1);
    linky.insertAfter(6, 5);
    System.out.println(linky);
    linky.insertBefore(0, 1);
    linky.insertBefore(4, 5);
    System.out.println(linky);
  }

  Node<T> head;
  Node<T> tail;

  public void insertBefore(T value, T target) throws Exception{
    if(head != null && head.value.equals(target)) {
      insert(value);
      return;
    }
    Node<T> current = head;;
    while(current != null){
      if(current.next.value.equals(target)){
       current.next = new Node(value, current.next); 
       return;
      }
      current = current.next;
    }
    throw new Exception("target value not in list");
  }

  public void insertAfter(T value, T target) throws Exception{
    Node<T> current = head;;
    while(current != null){
      if(current.value.equals(target)){
       current.next = new Node(value, current.next); 
       return;
      }
      current = current.next;
    }
    throw new Exception("target value not in list");
  }
  public void insert(T value){
    head = new Node<T>(value, head);
    if(tail == null) tail = head;
  }

  public void append(T value){
    if(tail == null) {
      insert(value);
      return;
    }
    tail.next = new Node<T>(value, null);
    tail = tail.next;
  }

  public boolean includes(T value){
    Node<T> current = head;
    while(current != null){
      if(current.value.equals(value)) return true;
      current = current.next;
    }
    return false;
  }

  public String toString(){
    StringBuilder stringified = new StringBuilder();
    Node<T> current = this.head;

    while(current != null){
      stringified.append("{" + current.value + "} -> ");
      current = current.next;
    }
    stringified.append("NULL");
    return stringified.toString();
  }

}

class Node<T> {
  T value;
  Node<T> next;

  public Node(T value, Node<T> next){
    this.value = value;
    this.next = next;
  }
}

File: c-sharp/CodeChallenge5/LinkedList.cs
namespace CodeChallenges
{
  public class LinkedList
  {
    public Node Head { get; set; }

    public LinkedList()
    {
      Head = null;

    }

    public void Insert(int value)
    {
      Node newNode = new Node(value);
      newNode.Next = Head;
      Head = newNode;
    }

    public bool Includes(int value)
    {
      Node current = Head;
      while (current != null)
      {
        if (current.Value == value)
          return true;
        current = current.Next;
      }
      return false;
    }

    public override string ToString()
    {
      if (Head == null)
        return "NULL";

      Node current = Head;
      string result = "";
      while (current != null)
      {
        result += $"{current.Value} -> ";
        current = current.Next;
      }
      result += "NULL";
      return result;
    }
  }

  public class Node
  {
    public int Value { get; set; }
    public Node Next { get; set; }

    public Node(int value)
    {
      Value = value;
      Next = null;
    }

    public static implicit operator Node(Node v)
    {
      throw new NotImplementedException();
    }
  }
}
